syncstorage/storage/sql/__init__.py

```diff
@@ -593,55 +670,91 @@ def delete_item(self, session, userid, collection, item):
-            total_affected += num_affected
+
+        for table_type in ("bso", "batch_upload_items"):
+            tabula = table_sets[table_type]
+            if table_type == "batch_upload_items":
+                # Tidy up the batch_uploads table first
+                if batches_purged <= max_per_loop:
+                    with self._get_or_create_session() as session:
+                        batches_purged = session.query("PURGE_BATCHES", {
+                            "grace": tabula["grace_period"],
+                            "maxitems": max_per_loop,
+                        })
+                    total_batches_purged += batches_purged
+
+            if not self.dbconnector.shard:
+                tables = set((table_type,))
+            else:
+                shard_func = self.dbconnector.get_bso_table
+                if table_type == "batch_upload_items":
+                    shard_func = self.dbconnector.get_batch_item_table
+                tables = set(shard_func(i).name
+                             for i in xrange(self.dbconnector.shardsize))
+                assert len(tables) == self.dbconnector.shardsize
+            # Purge each table in turn, summing rowcounts.
+            # We set an upper limit on the number of iterations, to avoid
+            # getting stuck indefinitely on a single table.
+            for table in sorted(tables):
+                logger.info("Purging expired items from %s", table)
+                num_iters = 1
+                num_affected = 0
+                with self._get_or_create_session() as session:
+                    rowcount = session.query(tabula["query"], {
+                        tabula["replace"]: table,
+                        "grace": tabula.get("grace_period", grace_period),
+                        "maxitems": max_per_loop,
+                    })
+                while rowcount > 0:
+                    num_affected += rowcount
+                    logger.debug("After %d iterations, %s items purged",
+                                 num_iters, num_affected)
+                    num_iters += 1
+                    if num_iters > 100:
+                        logger.debug("Too many iterations on %s, bailing outXS"
+                                     % table)
+                        is_incomplete = True
+                        break
+                    with self._get_or_create_session() as session:
+                        rowcount = session.query(tabula["query"], {
+                            tabula["replace"]: table,
+                            "grace": tabula.get("grace_period", grace_period),
+                            "maxitems": max_per_loop,
+                        })
+                logger.info("Purged %d expired items from %s",
+                            num_affected, table)
+                tabula["total_affected"] += num_affected
+
+        logger.info("Purged %d expired batches from "
+                    "batch_uploads" % total_batches_purged)
```

Comment:
Wow, this purge method wound up getting pretty gnarly.  I wonder if it would be clearer (if slightly more verbose) to make three separate methods `purge_expired_bsos`, `purge_expired_batches` `purge_expired_batch_items` rather than squish them all into a single control-flow here.